### PR TITLE
Block API: Add support for icons for block categories

### PR DESCRIPTION
**Pull request by tyxla**
## Description
This PR adds support for icons for block categories. In addition to supporting Dashicons, the idea is to introduce support for external components, functions and svg icons, in order to allow for third party plugins to be flexible with the icons for the categories they register. 

Since `<BlockIcon />` already supported all those kinds of methods to define an icon, we're abstracting the functionality to a separate component, and reusing it here and in `<BlockIcon />`.

## How has this been tested?
* Checkout this PR and build Gutenberg.
* Open the inserter in a post.
* Look at the "Common" category.
* Verify you can see the Dashicon to the right of the "Common" category name.
* Save at least one reusable block.
* Verify the reusable blocks category icon is displayed to the right of the category name.
* Verify that block categories without icons are unchanged.
* Verify existing block icons in the inserter are unchanged.
* Verify unit tests pass:
  * `BlockIcon` tests - `npm run test-unit packages/editor/src/components/block-icon/`
  * `Icon` tests - `npm run test-unit packages/components/src/icon/`
  * All unit tests - `npm run test-unit`
  * All tests - `npm test`

## Screenshots

Inserter with an icon on the "Common" block category:
~https://cldup.com/Eupr3BAnyc.png~

"Reusable" category in the inserter with one or more existing reusable blocks:

Inserter with a custom SVG icon on a custom category:
~https://cldup.com/etFPNTYIK3.png~

(see https://github.com/Automattic/wp-calypso/pull/27878 for testing it)

## Types of changes
* Introduced a new `<RawIcon />` component and exposed it.
* Added tests and documentation for `<RawIcon />`.
* Updated `<BlockIcon />` to use `<RawIcon />`.
* Updated the `PanelBody` to use `<RawIcon />` for block categories.
* Now passing category icon from inserter menu to `<PanelBody />`.
* Now registering the default categories with `null` icons (no icons).
* Now displaying the panel body icons on the right side of the panel body title.
* Moved and updated some tests.
* TO BE DELETED BEFORE MERGE - added a dashicon to the "Common" block category (https://github.com/WordPress/gutenberg/commit/204582b043c82ade5a3b24a2d9eb8130822dbf20).
* Renamed `<RawIcon />` to `<Icon />`.
* Made rendering of `Dashicon` default to a size of 20x20, while keeping 24x24 for the rest of the icon types.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 

## Notes

* Since we've added a dashicon to the "Common" block category (https://github.com/WordPress/gutenberg/commit/204582b043c82ade5a3b24a2d9eb8130822dbf20) this needs to be deleted before merging.

### CODE REVIEW

**Comment by ZebulanStanphill** [comment]
What if the text was always rendered with space on the left for an icon, so both the icons and the text would always line up?

**Comment by folletto** [comment]
For reference, this is an old mockup I prepared to avoid the issue with icon alignment: 

categories-blocks-branded

The icon on the right avoids having different indentations for the various categories, and at the same time optimizes space. :)

**Comment by mtias** [comment]
Note: we should also update:

image

**Comment by tyxla** [comment]
Good call, folks - I like the suggestion for the icon on the right side. Updated the PR, here are some screenshots:

Inserter with an icon on the "Common" block category:


"Reusable" category in the inserter with one or more existing reusable blocks:


Inserter with a custom SVG icon on a custom category:


This is ready for another look.

File: packages/components/src/raw-icon/index.js [hunk 1]
```diff
@@ -0,0 +1,35 @@
+/**
+ * WordPress dependencies
+ */
+import { cloneElement, createElement, Component, isValidElement } from '@wordpress/element';
+import { Dashicon, SVG } from '../';
+
+function RawIcon( { icon = null, size = 24, className } ) {
```
Comment (youknowriad) [inline packages/components/src/raw-icon/index.js]: I  suspect the default size property here might have some unexpected side effects on the different types of block icons and the different places where we render them. Mind a sanity check here @jasmussen or @chrisvanpatten as you know more about all the specificities.

File: packages/components/src/raw-icon/index.js [hunk 2]
```diff
@@ -0,0 +1,35 @@
+/**
```
Comment (youknowriad) [inline packages/components/src/raw-icon/index.js]: 👍 Nice little component. Why not just `Icon`, Thoughts on the name @aduth @mcsf

File: packages/components/src/panel/body.js [hunk 3]
```diff
@@ -61,8 +61,8 @@ class PanelBody extends Component {
 									<G><Path d="M7.41,8.59L12,13.17l4.59-4.58L18,10l-6,6l-6-6L7.41,8.59z" /></G>
 								</SVG>
 							}
-							{ icon && <Dashicon icon={ icon } className="components-panel__icon" /> }
 							{ title }
+							{ icon && <RawIcon icon={ icon } className="components-panel__icon" size={ 20 } /> }
```
Comment (youknowriad) [inline packages/components/src/panel/body.js]: What's the impact of this change on other usage of `PanelBody` with icons? Are we ok with all panel body icons being moved to the right or should it be specific to the inserter?

File: packages/components/src/raw-icon/index.js (hunk 2 above)
Comment (aduth) [reply packages/components/src/raw-icon/index.js]: My expectations from the name and description don't really follow into the implementations. If it's "without styles", why do we apply a default width? And an unstyled variation seems more appropriate as a prop of a default component type, e.g. `<Icon isUnstyled />`.

**Review by youknowriad — COMMENTED** [review]
At the moment custom categories are provided using a backend filter IIRC cc @gziolo, how do we provide SVGs from the backend? Do we have the corresponding JavaScript API to add these categories?

File: packages/components/src/raw-icon/index.js (hunk 1 above)
Comment (jasmussen) [reply packages/components/src/raw-icon/index.js]: Interesting timing, I'm actually trying to fix a regression in https://github.com/WordPress/gutenberg/pull/10938#issuecomment-432220310 related to this. 

24 _should_ be default, but it's okay if an SVG needs to override that and show a 20x20 icon. 

If 24 is not default, then that's a bug, as is discussed in that link.

File: packages/components/src/raw-icon/index.js (hunk 2 above)
Comment (tyxla) [reply packages/components/src/raw-icon/index.js]: Well, I consider the width an attribute, and not a style 🤷‍♂️ 

And regarding the component name - `Icon` felt a little too broad, and I'd probably expect an icon component to be more flexible in terms of how the icon could _look_. So I'd instead expect to have an `Icon` component that could use the `RawIcon`, but add on top of it. 

What do y'all think?

File: packages/components/src/panel/body.js (hunk 3 above)
Comment (tyxla) [reply packages/components/src/panel/body.js]: Last time I checked, the only usages were in the inserter, and this PR has verified them all. 

Whether we're planning to use the `PanelBody` for something else, and icons would make sense on the right in these cases - perhaps @mtias @jasmussen would know better.

File: packages/components/src/raw-icon/index.js (hunk 1 above)
Comment (youknowriad) [reply packages/components/src/raw-icon/index.js]: It looks like we were explicitly showing Dashicons as `20x20` while this PR makes them `24x24` ?

File: packages/components/src/panel/body.js (hunk 3 above)
Comment (youknowriad) [reply packages/components/src/panel/body.js]: It looks like it was introduced explicitly for this particular usage (Inserter), so yes, I think I'm fine with the change.

Comment (mtias) [reply packages/components/src/panel/body.js]: Yes, it was added for the reusable block feature.

File: packages/components/src/raw-icon/index.js (hunk 2 above)
Comment (tyxla) [reply packages/components/src/raw-icon/index.js]: I'm good with changing to `UnstyledIcon`, @aduth @youknowriad does that make sense to you?

Comment (aduth) [reply packages/components/src/raw-icon/index.js]: >I'd probably expect an icon component to be more flexible in terms of how the icon could look.

Can you elaborate on what you mean by this? I'm not sure I follow, and am having a hard time understanding what specifically we're implying in these multiple layers of icon components.

Comment (youknowriad) [reply packages/components/src/raw-icon/index.js]: I personally still have a small preference for just  `Icon`, if we need to make a distinction between styled and unstyled later, we can add it as a prop.

Comment (tyxla) [reply packages/components/src/raw-icon/index.js]: I've been thinking and started gathering a list of what `<Icon />` would provide more than `<UnstyledIcon />`, but I didn't get too far with it. 

I guess I'm good with renaming this one to `<Icon />`, it seems @mcsf and @youknowriad are good with `<Icon />`, @aduth does that sound well to you?

Comment (aduth) [reply packages/components/src/raw-icon/index.js]: Works for me. 👍

Comment (tyxla) [reply packages/components/src/raw-icon/index.js]: Cool, thanks 👍 

Renamed `<RawIcon />` to `<Icon />` in https://github.com/WordPress/gutenberg/pull/10651/commits/3486d0bcc6d2926f453e7b0586e378f93d2c4562

I'd appreciate another look when y'all get a chance.

File: packages/block-library/src/embed/test/__snapshots__/index.js.snap [hunk 4]
```diff
@@ -14,10 +14,10 @@ exports[`core/embed block edit matches snapshot 1`] = `
         aria-hidden="true"
         class="dashicon dashicons-embed-generic"
         focusable="false"
-        height="20"
+        height="24"
```
Comment (youknowriad) [inline packages/block-library/src/embed/test/__snapshots__/index.js.snap]: I think this is an issue actually. As outlined here https://github.com/WordPress/gutenberg/pull/10651#discussion_r227423954 Dashicons should be 20x20 by default.

Comment (tyxla) [reply packages/block-library/src/embed/test/__snapshots__/index.js.snap]: Oh, that's a fair point. I'll reflect this in the component then, and will add some tests too.

Comment (tyxla) [reply packages/block-library/src/embed/test/__snapshots__/index.js.snap]: Updated in https://github.com/WordPress/gutenberg/pull/10651/commits/c7959cfc3308034bfa1763cc35317f065cfbcb7e, thanks for the suggestion!

**Review by youknowriad — APPROVED** [review]
I didn't test but looks good to me code-wise :)

**Comment by youknowriad** [comment]
Tested and works well.

**Comment by designsimply** [comment]
I tested with WordPress 4.9.8 and Gutenberg 4.2.0-rc.1 and didn't see a Dashicon to the right of the "Common" category name. Is this because it's something that needs to be opted into using custom code? Note: I followed the testing steps quite literally and it's possible I'm missing something.

**Comment by tyxla** [comment]
@designsimply - it's expected that you don't see it. Adding the Common category icon was for testing purposes for this PR, and even though we merged it with this PR, we reverted it in #11191.

**Comment by designsimply** [comment]
Aha. I see. Thank you for the clarification!

**Pull request merged**